libpkg/pkg_jobs.c

```diff
@@ -1098,14 +1098,20 @@ pkg_jobs_need_upgrade(struct pkg *rp, struct pkg *lp)
 		ret2 = pkg_deps(lp, &ld);
 		if (ret1 != ret2) {
 			free(rp->reason);
-			rp->reason = strdup("direct dependency changed");
+			asprintf(&rp->reason, "direct dependency changed: %s", 
+			    rd->name);
+			if (rp->reason == NULL)
+				rp->reason = strdup("direct dependency changed");
```

Comment:
As I've said privately, this `strdup` is not needed here and the whole check should be just switched to `assert(rp->reason != NULL)` that follows `asprintf`.